cli/azd/pkg/containerregistry/remote_build.go

```diff
@@ -0,0 +1,202 @@
+	ctx, span := tracing.Start(ctx, events.RemoteBuildUploadEvent)
+
+	res, err := func() (armcontainerregistry.SourceUploadDefinition, error) {
+		cred, err := r.credentialProvider.CredentialForSubscription(ctx, subscriptionID)
+		if err != nil {
+			return armcontainerregistry.SourceUploadDefinition{}, err
+		}
+
+		regClient, err := armcontainerregistry.NewRegistriesClient(subscriptionID, cred, r.armClientOptions)
+		if err != nil {
+			return armcontainerregistry.SourceUploadDefinition{}, err
+		}
+
+		sourceUploadRes, err := regClient.GetBuildSourceUploadURL(ctx, resourceGroupName, registryName, nil)
+		if err != nil {
+			return armcontainerregistry.SourceUploadDefinition{}, err
+		}
+
+		blobClient, err := blockblob.NewClientWithNoCredential(*sourceUploadRes.UploadURL, nil)
+		if err != nil {
+			return armcontainerregistry.SourceUploadDefinition{}, err
+		}
+
+		dockerContext, err := os.Open(buildSourcePath)
+		if err != nil {
+			return armcontainerregistry.SourceUploadDefinition{}, err
+		}
+		defer dockerContext.Close()
+
+		_, err = blobClient.UploadFile(context.Background(), dockerContext, nil)
+		if err != nil {
+			return armcontainerregistry.SourceUploadDefinition{}, err
+		}
+
+		return sourceUploadRes.SourceUploadDefinition, nil
+	}()
+
+	span.EndWithStatus(err)
+	return res, err
+}
+
+// RunDockerBuildRequestWithLogs initiates a remote build on the specified registry and streams the logs to the provided
+// writer.
+func (r *RemoteBuildManager) RunDockerBuildRequestWithLogs(
+	ctx context.Context,
+	subscriptionID, resourceGroupName, registryName string,
+	buildRequest *armcontainerregistry.DockerBuildRequest,
+	writer io.Writer,
+) error {
+	ctx, span := tracing.Start(ctx, events.RemoteBuildEvent)
+
+	err := func() error {
+		cred, err := r.credentialProvider.CredentialForSubscription(ctx, subscriptionID)
+		if err != nil {
+			return err
+		}
+
+		regClient, err := armcontainerregistry.NewRegistriesClient(subscriptionID, cred, r.armClientOptions)
+		if err != nil {
+			return err
+		}
+
+		runPoller, err := regClient.BeginScheduleRun(ctx, resourceGroupName, registryName, buildRequest, nil)
+		if err != nil {
+			return err
+		}
+
+		resp, err := runPoller.Poll(ctx)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+		var runResp armcontainerregistry.RegistriesClientScheduleRunResponse
+		if err := json.NewDecoder(resp.Body).Decode(&runResp); err != nil {
+			return err
+		}
+
+		runClient, err := armcontainerregistry.NewRunsClient(subscriptionID, cred, r.armClientOptions)
+		if err != nil {
+			return err
+		}
+
+		logRes, err := runClient.GetLogSasURL(ctx, resourceGroupName, registryName, *runResp.Properties.RunID, nil)
+		if err != nil {
+			return err
+		}
+
+		logBlobClient, err := blockblob.NewClientWithNoCredential(*logRes.LogLink, &blockblob.ClientOptions{
+			ClientOptions: r.armClientOptions.ClientOptions,
+		})
+		if err != nil {
+			return err
+		}
+
+		err = streamLogs(ctx, logBlobClient, writer)
```

Comment:
I'll dig into this more and also compare what happens with `az acr build`.

Reply:
I looked into this more and compared with with `az acr build` does.  I suspect the error that you saw here was actually a race and that the log file would have become available.  I repeated your experiment and saw that the log was produced (and included the build failure) but `azd` was not checking the status of the operation after it had completed so even if the container build failed we'd continue along (the overall deploy operation would fail because the image we'd try to update to didn't exist).

I modified the log pumping logic to do a limited number of retries on 404s (on the assumption that the file the service told us about will soon exist) and then made updates to check the job status after the log file is marked as complete to see if the overall operation worked or not.